File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/F-Алхимия/solution.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <queue>

using namespace std;

void solution() {
    /*
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V|)
    */
    int m;
    scanf("%d", &m);
    cin.ignore();
    unordered_map<string, vector<string>> adjList;
    for (int i = 0; i < m; ++i) {
        string convertation;
        getline(cin, convertation);
        int x = convertation.find(" -> ");
        string source = convertation.substr(0, x);
        string destination = convertation.substr(x + 4);
        adjList[destination].push_back(source);
    }
    string start, target;
    cin >> start >> target;
    unordered_set<string> seen;
    seen.insert(target);
    queue<pair<string, int>> q;
    q.push({target, 0});
    while (!q.empty()) {
        string vertex = q.front().first;
        int path = q.front().second;
        q.pop();
        if (vertex == start) {
            printf("%d\n", path);
            return;
        }
        for (string& neighbor : adjList[vertex]) {
            if (!seen.count(neighbor)) {
                seen.insert(neighbor);
                q.push({neighbor, path + 1});
            }
        }
    }
    printf("-1\n");
    
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/F-Алхимия/solution.py
import sys
from collections import deque


def solution() -> None:
    """
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V|)
    """
    m: int = int(sys.stdin.readline().rstrip())
    adj_list: dict[str, list[str]] = dict()
    for _ in range(m):
        convertation: str = sys.stdin.readline().rstrip()
        source, destination = convertation.split(' -> ')
        if destination not in adj_list:
            adj_list[destination] = []
        adj_list[destination].append(source)
    start: str = sys.stdin.readline().rstrip()
    target: str = sys.stdin.readline().rstrip()
    seen: set[str] = set()
    q: list[tuple[str, int]] = deque()
    q.append((target, 0))
    while q:
        vertex, path = q.popleft()
        if vertex == start:
            sys.stdout.write('{}\n'.format(path))
            return
        for neighbor in adj_list.get(vertex, []):
            if neighbor not in seen:
                seen.add(neighbor)
                q.append((neighbor, path + 1))
    sys.stdout.write('{}\n'.format(-1))


if __name__ == '__main__':
    solution()